tests/test_app/TestApp/Database/Type/PointType.php

```diff
@@ -0,0 +1,97 @@
+<?php
+declare(strict_types=1);
+
+namespace TestApp\Database\Type;
+
+use Cake\Database\DriverInterface;
+use Cake\Database\Expression\FunctionExpression;
+use Cake\Database\ExpressionInterface;
+use Cake\Database\Schema\TableSchemaInterface;
+use Cake\Database\Type\BaseType;
+use Cake\Database\Type\ExpressionTypeInterface;
+use Cake\Database\Type\SchemaAwareTypeInterface;
+use Cake\Database\Type\SelectExpressionTypeInterface;
+use TestApp\Database\Point;
+
+class PointType extends BaseType implements ExpressionTypeInterface, SelectExpressionTypeInterface, SchemaAwareTypeInterface
+{
+    public function toPHP($value, DriverInterface $d)
+    {
+        return Point::fromGeoJSON(json_decode($value, true));
+    }
+
+    public function marshal($value)
+    {
+        if (is_string($value)) {
+            $value = explode(',', $value);
+        }
+
+        if (is_array($value)) {
+            return new Point($value[0], $value[1]);
+        }
+
+        return null;
+    }
+
+    public function toExpression($value): ExpressionInterface
+    {
+        if ($value instanceof Point) {
+            return new FunctionExpression('POINT', [$value->lat(), $value->long()]);
+        }
+
+        if (is_array($value)) {
+            return new FunctionExpression('POINT', [$value[0], $value[1]]);
+        }
+
+        throw new \InvalidArgumentException(sprintf(
+            'The `$value` argument must be an instance of `\%s`, or an array, `%s` given.',
+            Point::class,
+            getTypeName($value)
+        ));
+    }
+
+    public function toSelectExpression($field): ExpressionInterface
+    {
+        if (is_string($field)) {
+            $args = [$field => 'identifier'];
+        } else {
+            $args = [$field];
+        }
+
+        return new FunctionExpression('ST_AsGeoJSON', $args);
+    }
+
+    public function toDatabase($value, DriverInterface $driver)
+    {
+        return $value;
+    }
+
+    public function getSchemaColumnSql(TableSchemaInterface $schema, string $column, DriverInterface $driver): ?string
```

Comment:
Would types use branching conditions for each DB platform they support?

Reply:
They don't have to, but they could if they wanted to.

The types could evaluate the driver (and aspects of the underlying DBMS, like the version) and return either driver specific `SQL`, or `null` if they don't want to handle a specific dialect and fall back to letting CakePHP try to handle it.

Reply:
I thought about passing the driver to the expression generation methods too, so that they could generate driver specific expressions, but I was unsure about it.

Reply:
I think keeping driver specific conditions here is fine for now. I feel like custom types will likely end up being specific to a single dialect so that users can leverage proprietary extensions in that dialect.

I would keep expression objects unaware of the dialect they are being used for as much as possible. For built in expressions we use the `_expressionTranslators()` and `_xxQueryTranslator()` hooks to provide dialect specific expression transforms. If we need a more elaborate transform system for custom expressions we can extend that system.